### PR TITLE
Bag auditor checks ingest type before assigning a version

### PR DESCRIPTION
**Pull request by alexwlchan**
Closes wellcometrust/platform#3657, and possibly wellcometrust/platform#3499?

It includes some new user-facing messages in the ingests response:

> Unable to find an external identifier
>
> This bag has never been ingested before, but was sent with ingestType update
>
> This bag has already been ingested, but was sent with ingestType create
>
> Another version of this bag was ingested at ${e.stored}, which is newer than the current ingest ${e.request}

I discovered wellcometrust/platform#3693 while working on this.

### CODE REVIEW

**Comment by jtweed** [comment]
What about something like this for the messages:

- Unable to ingest bag. An external identifier was not found in the bag info.
- Unable to ingest bag. Another version of this bag was ingested at $PREVIOUS_INGEST_DATE, which is newer than the current ingest $CURRENT_INGEST_DATE.
- Unable to ingest new bag. A bag with the supplied external identifier already exists in this space.
- Unable to update existing bag. A bag with the supplied external identifier does not exist in this space.

**Comment by alexwlchan** [comment]
@jtweed I’m picking nits, but “external identifier was not found in the bag info” implies that were able to find the bag info. This step could also fail if the bag info is just missing. Do we care about that distinction?

**Comment by jtweed** [comment]
If the bag info itself is missing, that would ideally be a distinct error. If we can’t tell the difference, then I think this message is still correct, as we couldn’t find an id in the bag info, as there was no bag info.

**Comment by alexwlchan** [comment]
> If the bag info itself is missing, that would ideally be a distinct error. If we can’t tell the difference, then I think this message is still correct, as we couldn’t find an id in the bag info, as there was no bag info.

Not currently; I'll change the error messages here and open a new ticket to make it a distinct error, because I expect it'll need a moderate refactor.

File: bag_auditor/src/main/scala/uk/ac/wellcome/platform/storage/bagauditor/versioning/VersionPicker.scala [hunk 1]
```diff
@@ -3,57 +3,87 @@ package uk.ac.wellcome.platform.storage.bagauditor.versioning
...
       }
-      .map {
-        case Right(version) =>
-          checkVersionIsAllowed(ingestType, assignedVersion = version)
-          version
 
-        case Left(FailedProcess(_, err)) => throw err
-        case Left(err)                   => throw new RuntimeException(s"Locking error: $err")
-      }
+    // This is a double Either: the outer Either defines the result of the locking
```
Comment (kenoir) [inline bag_auditor/src/main/scala/uk/ac/wellcome/platform/storage/bagauditor/versioning/VersionPicker.scala]: https://typelevel.org/cats/datatypes/nested.html

Might come in handy next time!

File: bag_auditor/src/main/scala/uk/ac/wellcome/platform/storage/bagauditor/versioning/VersionPicker.scala [hunk 2]
```diff
@@ -3,57 +3,87 @@ package uk.ac.wellcome.platform.storage.bagauditor.versioning
...
+      Left(IngestTypeCreateForExistingBag())
     } else if (ingestType == UpdateIngestType && assignedVersion == 1) {
-      throw new IllegalVersionAssignment(
-        "Ingest type 'update' is not allowed unless a bag already exists")
+      Left(IngestTypeUpdateForNewBag())
+    } else {
+      Right(assignedVersion)
+    }
+
+  // Annoyingly, cats doesn't seem to provide an Implicit for the Id monad, so we have
```
Comment (kenoir) [inline bag_auditor/src/main/scala/uk/ac/wellcome/platform/storage/bagauditor/versioning/VersionPicker.scala]: I think it does? Does `import cats._` not work?

File: bag_auditor/src/main/scala/uk/ac/wellcome/platform/storage/bagauditor/versioning/VersionPicker.scala (hunk 1 above)
Comment (alexwlchan) [reply bag_auditor/src/main/scala/uk/ac/wellcome/platform/storage/bagauditor/versioning/VersionPicker.scala]: Next time!

**Pull request merged**